p2p/http/libp2phttp.go

```
@@ -0,0 +1,790 @@
+func normalizeHTTPMultiaddr(addr ma.Multiaddr) (ma.Multiaddr, bool) {
+	isHTTPMultiaddr := false
+	beforeHTTPS, afterIncludingHTTPS := ma.SplitFunc(addr, func(c ma.Component) bool {
+		if c.Protocol().Code == ma.P_HTTP {
+			isHTTPMultiaddr = true
+		}
+
+		if c.Protocol().Code == ma.P_HTTPS {
+			isHTTPMultiaddr = true
+			return true
+		}
+		return false
+	})
+
+	if afterIncludingHTTPS == nil {
+		// No HTTPS component, just return the original
+		return addr, isHTTPMultiaddr
+	}
+
+	_, afterHTTPS := ma.SplitFirst(afterIncludingHTTPS)
+	if afterHTTPS == nil {
+		return ma.Join(beforeHTTPS, tlsComponent, httpComponent), isHTTPMultiaddr
+	}
+
+	return ma.Join(beforeHTTPS, tlsComponent, httpComponent, afterHTTPS), isHTTPMultiaddr
+}
+
+// ProtocolPathPrefix looks up the protocol path in the well-known mapping and
+// returns it. Will only store the peer's protocol mapping if the server ID is
+// provided.
+func (h *Host) getAndStorePeerMetadata(roundtripper http.RoundTripper, server peer.ID) (PeerMeta, error) {
+	if h.peerMetadata == nil {
+		h.peerMetadata = newPeerMetadataCache()
+	}
+	if meta, ok := h.peerMetadata.Get(server); server != "" && ok {
+		return meta, nil
+	}
+
+	req, err := http.NewRequest("GET", "/.well-known/libp2p", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+
+	client := http.Client{Transport: roundtripper}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body := [peerMetadataLimit]byte{}
+	bytesRead := 0
+	for {
+		n, err := resp.Body.Read(body[bytesRead:])
+		bytesRead += n
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if bytesRead >= peerMetadataLimit {
+			return nil, fmt.Errorf("peer metadata too large")
+		}
+	}
+
+	meta := PeerMeta{}
+	json.Unmarshal(body[:bytesRead], &meta)
```

Ignored error return value?